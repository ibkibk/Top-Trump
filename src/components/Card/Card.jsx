import React, { Component } from "react";
import styles from "./Card.module.scss";
import PropTypes from "prop-types";

// Class containing JSX & methods associated with the card component

class Card extends Component {
  // Function check the length of the card's description & either returns it whole or shortened
  getDescription = () => {
    const description = this.props.cardData.description;
    return description.length > 120
      ? description.substring(0, 120) + "..."
      : description;
  };

  // object containing any properties related to component state
  state = {
    isFaceUp: true
  };

  // setState allows us to update state indirectly
  handleClick = () => {
    this.setState({
      isFaceUp: !this.state.isFaceUp
    });
  };

  /*
    render function describes what the UI will look like using JSX.
  */

  render() {
    const rotateStyle = this.state.isFaceUp ? "" : styles.cardRotated;
    return (
      <div className={styles.scene}>
        <article
          onClick={this.handleClick}
          className={`${styles.card} ${rotateStyle}`}
        >
          <div className={styles.cardFront}>
            <h3>{this.props.cardData.name}</h3>
            <div className={styles.topSection}>
              <section className={styles.description}>
                {this.getDescription()}
              </section>
              <img
                className={styles.picture}
                alt={this.props.cardData.name + "'s face"}
                src={this.props.cardData.picture}
              />
            </div>
            <table className={styles.skills}>
              <tbody>
                <tr>
                  <th>Loudness</th>
                  <td>{this.props.cardData.skills.loudness}</td>
                </tr>
                <tr>
                  <th>Greed</th>
                  <td>{this.props.cardData.skills.greed}</td>
                </tr>
                <tr>
                  <th>Attitude</th>
                  <td>{this.props.cardData.skills.attitude}</td>
                </tr>
                <tr>
                  <th>Sass</th>
                  <td>{this.props.cardData.skills.sass}</td>
                </tr>
                <tr>
                  <th>MVP</th>
                  <td>{this.props.cardData.skills.mvp}</td>
                </tr>
              </tbody>
            </table>
          </div>
          <div className={styles.cardBack} />
        </article>
      </div>
    );
  }
}

// Using propTypes to define the data types for the 'skills' object

const skillsTypes = PropTypes.shape({
  loudness: PropTypes.number.isRequired,
  greed: PropTypes.number.isRequired,
  attitude: PropTypes.number.isRequired,
  sass: PropTypes.number.isRequired,
  mvp: PropTypes.number.isRequired
});

// Using propTypes to define the data types for
// the props object passed to the card component

Card.propTypes = {
  cardData: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
    skills: skillsTypes
  })
};

export default Card;
