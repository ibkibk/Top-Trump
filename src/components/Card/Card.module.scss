/* playing cards ratio is 2.5:3.5 */

/* SASS variable setup */
$card-width: 265px;
$card-height: $card-width * 3.5 / 2.5;
$picture-width: 100px;
$picture-height: $picture-width * 3.5 / 2.5;

.scene {
  width: $card-width;
  min-width: $card-width;
  height: $card-height;
  perspective: 600px;
  margin: 10px;
}

.card {
  text-align: center;
  width: 100%;
  height: 100%;
  position: relative;
  transition: transform 1s;
  transform-style: preserve-3d;

  h3 {
    margin-bottom: 10px;
  }
}

%card {
  border: 8px solid black;
  background-size: cover;
  border-radius: 15px;
  background-repeat: no-repeat, repeat;
  position: absolute;
  height: calc(100% - 16px);
  width: calc(100% - 16px);
  backface-visibility: hidden;
}

.cardFront {
  @extend %card;
  background-image: url(../../static/images/card.png);
}

.cardBack {
  @extend %card;
  background-image: url(../../static/images/card-back.png);
  transform: rotateY(180deg);
}

.cardRotated {
  transform: rotateY(180deg);
}

.picture {
  width: $picture-width;
  height: $picture-height;
  border: 1px solid black;
}

.topSection {
  display: flex;
  justify-content: space-around;
}

.description {
  width: $picture-width;
  height: $picture-height;
  font-size: small;
}

.skills {
  margin: 5px auto 5px auto;
  width: 100%;
  border-spacing: 5px 5px;
  td,
  th {
    background-color: black;
    color: white;
    border: 1px solid white;
    padding: 2px 15px 2px 15px;
    text-align: right;
  }
  td {
    border-top-left-radius: 25px;
  }
  th {
    border-bottom-right-radius: 25px;
  }
}
