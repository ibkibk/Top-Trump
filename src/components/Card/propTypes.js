import PropTypes from "prop-types";

// Using propTypes to define the data types for the 'skills' object

export const skillsTypes = PropTypes.objectOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number]));

// Using propTypes to define the data types for
// the props object passed to the card component

export const cardTypes = {
  cardData: PropTypes.shape({
    selectCard: PropTypes.func,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
    skills: skillsTypes
  })
};
