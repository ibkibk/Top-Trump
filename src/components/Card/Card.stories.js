import React from "react";
import Card from "./Card";
import craig from "../../static/images/craig.png";

export default {
  title: "Card"
};

const mockData = {
  description:
    "Craig is Head of Technology for _nology. He loves rock climbing and solving problems. The only problem craig hasn't been able to solve is how to climb while leaving his ego on the ground.",
  name: "Craig Livings",
  picture: craig,
  role: "Trainer",
  skills: { loudness: 5, greed: 3, attitude: 8, sass: 7, mvp: 9 }
};

export const card = () => <Card cardData={mockData} />;
