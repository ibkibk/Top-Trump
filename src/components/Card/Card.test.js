import React from "react";
import Card from "./Card";
import craig from "../../static/images/craig.png";

// mount function allows us to render a component instance - including any children
import { mount } from "enzyme";

// mock data object to be passed as props into test component instance
const mockData = {
  description:
    "Craig is Head of Technology for _nology. He loves rock climbing and solving problems. The only problem craig hasn't been able to solve is how to climb while leaving his ego on the ground.",
  name: "Craig Livings",
  picture: craig,
  role: "Trainer",
  skills: { loudness: 5, greed: 3, attitude: 8, sass: 7, mvp: 9 }
};

describe("CardComponent Tests", () => {
  let component;

  // beforeEach() provides setup to run prior to every test

  beforeEach(() => {
    component = mount(<Card cardData={mockData} />);
  });

  it("should contain a HTML image tag", () => {
    expect(component.find("img").length).toEqual(1);
  });

  it("should receive data passed as props from parent", () => {
    expect(component.props().cardData).toEqual(mockData);
  });

  it("should render data passed in as props", () => {
    expect(component.text()).toContain(mockData.name);
    expect(component.text()).toContain(mockData.description.substring(0, 175));
    expect(component.find(mockData.picture)).toBeTruthy();
  });

  it("should match the snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
