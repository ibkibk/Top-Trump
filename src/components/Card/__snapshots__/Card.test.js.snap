// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardComponent Tests should match the snapshot 1`] = `
<Card
  cardData={
    Object {
      "description": "Craig is Head of Technology for _nology. He loves rock climbing and solving problems. The only problem craig hasn't been able to solve is how to climb while leaving his ego on the ground.",
      "name": "Craig Livings",
      "picture": "craig.png",
      "role": "Trainer",
      "skills": Object {
        "attitude": 8,
        "greed": 3,
        "loudness": 5,
        "mvp": 9,
        "sass": 7,
      },
    }
  }
>
  <article
    className="card"
  >
    <h3>
      Craig Livings
    </h3>
    <div
      className="topSection"
    >
      <section
        className="description"
      >
        Craig is Head of Technology for _nology. He loves rock climbing and solving problems. The only problem craig hasn't been able to solve is how to climb while leaving his ego on...
      </section>
      <img
        alt="Craig Livings's face"
        className="picture"
        src="craig.png"
      />
    </div>
    <table
      className="skills"
    >
      <tbody>
        <tr>
          <th>
            Loudness
          </th>
          <td>
            5
          </td>
        </tr>
        <tr>
          <th>
            Greed
          </th>
          <td>
            3
          </td>
        </tr>
        <tr>
          <th>
            Attitude
          </th>
          <td>
            8
          </td>
        </tr>
        <tr>
          <th>
            Sass
          </th>
          <td>
            7
          </td>
        </tr>
        <tr>
          <th>
            MVP
          </th>
          <td>
            9
          </td>
        </tr>
      </tbody>
    </table>
  </article>
</Card>
`;
